feature_engine/timeseries/forecasting/window_features.py

```diff
@@ -0,0 +1,333 @@
+
+        {feature_names_in_}
+
+        {n_features_in_}
+
+        Methods
+        -------
+        {fit}
+
+        transform:
+            Add lag features.
+
+        {fit_transform}
+
+        See Also
+        --------
+        pandas.shift
+        """
+
+    def __init__(
+            self,
+            variables: List[str] = None,
+            window: Union[str, int] = 1,
+            function: Callable = np.mean,
+            periods: int = 1,
+            freq: str = None,
+            sort_index: bool = True,
+            missing_values: str = "raise",
+            drop_original: bool = False,
+    ) -> None:
+
+        if not (isinstance(window, int) or isinstance(window, str)) or isinstance(window, bool):
+            raise ValueError(
+                f"window must be a string or integer. Got {window} instead."
+            )
+
+        if isinstance(periods, int) and periods > 0 and not isinstance(periods, bool):
+            self.periods = periods
+        else:
+            raise ValueError(
+                f"periods must be a positive integer. Got {periods} instead."
+            )
+
+        if function not in (np.mean, np.std, np.median, np.min, np.max, np.sum):
+            raise ValueError(
+                f"function must be np.mean, np.std, np.median, np.min, np.max, or np.sum. "
+                f"Got {function} instead."
+            )
+        if periods is None and not isinstance(freq, str):
+            raise ValueError(
+                f"freq must be a string. Got {freq} instead."
+            )
+
+        if missing_values not in ["raise", "ignore"]:
+            raise ValueError(
+                "missing_values takes only values 'raise' or 'ignore'. "
+                f"Got {missing_values} instead."
+            )
+
+        if not isinstance(drop_original, bool):
+            raise ValueError(
+                "drop_original takes only boolean values True and False. "
+                f"Got {drop_original} instead."
+            )
+
+        self.variables = _check_input_parameter_variables(variables)
+        self.window = window
+        self.function = function
+        self.freq = freq
+        self.sort_index = sort_index
+        self.missing_values = missing_values
+        self.drop_original = drop_original
+
+    def fit(self, X: pd.DataFrame, y: Optional[pd.Series] = None):
+        """
+        This transformer does not learn parameters.
+
+        Parameters
+        ----------
+        X: pandas dataframe of shape = [n_samples, n_features]
+            The training dataset.
+
+        y: pandas Series, default=None
+            y is not needed in this transformer. You can pass None or y.
+        """
+        # check input dataframe
+        X = _is_dataframe(X)
+
+        # We need the dataframes to have unique values in the index and no missing data.
+        # Otherwise, when we merge the window features we will duplicate rows.
+
+        if X.index.isnull().sum() > 0:
```

Comment:
Is there any reason for doing this over this:

```
if X.index.isnull().any():
```

if not I believe my suggestion is the preferred method for such checks.

Reply:
fixed.

Reply:
> few questions regarding the co

feature_engine checks are not failing for me in this pr https://github.com/Morgan-Sell/feature_engine/pull/9

sklearn check_estimators tests are failing. And not sure why yet.